### PR TITLE
feat(filter): add possibility to filter on smart field

### PR DESCRIPTION
**Pull request by VincentMolinie**
## Pull Request checklist:

- [ ] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [ ] Create automatic tests
- [ ] No automatic tests failures
- [ ] Test manually the implemented changes
- [ ] Review my own code (indentation, syntax, style, simplicity, readability)
- [ ] Wonder if you can improve the existing code

## Snippet for fullname field
```ruby
class Forest::Customer
  include  ForestLiana::Collection

  collection :Customer

  filter_fullname = lambda do |condition, where|
    fullname = condition['value']
    fullname_array = fullname.split
    firstname = fullname_array.first
    lastname = fullname_array.last

    where_array = where.split
    operator = where_array[1]
    value = where_array.drop(2).join(" ")
    firstname_value = value.sub fullname, firstname
    lastname_value = value.sub fullname, lastname


    "(firstname #{operator} #{firstname_value} AND lastname #{operator} #{lastname_value})"
  end

  field :fullname, type: 'String', is_read_only: false, is_required: true, is_filterable: true, filter: filter_fullname do
    "#{object.firstname} #{object.lastname}"
  end
end

```

### CODE REVIEW

File: app/services/forest_liana/filters_parser.rb [hunk 1]
```diff
@@ -47,28 +47,47 @@ def parse_aggregation(node)
...
+      where = parse_condition_without_smart_field(condition)
+
+      if SchemaUtils.is_smart_field(@resource, field)
+        schema = ForestLiana.schema_for_resource(@resource)
+        field_schema = schema.fields.find do |field|
+          field.try(:name) == field_name
+        end
+
+        unless field_schema.try(:[], :filter)
+          raise ForestLiana::Errors::ExpectedError.new(500, "Server Error", "method filter on smart field '#{field_name}' not found")
```
Comment (larcin) [inline app/services/forest_liana/filters_parser.rb:59]: [MISC] You're raising an expected error but it's a 500, a bit counter intuitive don't you think?

File: app/services/forest_liana/filters_parser.rb [hunk 2]
```diff
@@ -47,28 +47,47 @@ def parse_aggregation(node)
...
     def parse_condition(condition)
+      where = parse_condition_without_smart_field(condition)
+
+      if SchemaUtils.is_smart_field(@resource, field)
+        schema = ForestLiana.schema_for_resource(@resource)
+        field_schema = schema.fields.find do |field|
+          field.try(:name) == field_name
+        end
+
+        unless field_schema.try(:[], :filter)
```
Comment (larcin) [inline app/services/forest_liana/filters_parser.rb:58]: [MISC] I don't get it, you check for filter existence but don't use it afterward... 🤔

File: app/services/forest_liana/filters_parser.rb [hunk 3]
```diff
@@ -165,6 +184,9 @@ def is_belongs_to(field)
       field.include?(':')
     end
 
+    def is_smart_field(field)
+    end
```
Comment (larcin) [inline app/services/forest_liana/filters_parser.rb:188]: [MISC] Defined only to be overridden right?

**Review by larcin — CHANGES_REQUESTED** [review]
It seems that you made travis and CC angry :)

File: app/services/forest_liana/filters_parser.rb (hunk 3 above)
Comment (VincentMolinie) [reply app/services/forest_liana/filters_parser.rb:188]: Oopsy nop

Comment (arnaudbesnier) [reply app/services/forest_liana/filters_parser.rb:188]: Friday morning != Friday afternoon guys 🙏

File: app/services/forest_liana/filters_parser.rb (hunk 1 above)
Comment (VincentMolinie) [reply app/services/forest_liana/filters_parser.rb:59]: The name of the error was bad. I renamed it to HTTPError

File: app/services/forest_liana/filters_parser.rb (hunk 2 above)
Comment (VincentMolinie) [reply app/services/forest_liana/filters_parser.rb:58]: Mistake I wrote `:search` while it was actually `:filter` 😅

File: spec/services/forest_liana/resources_getter_spec.rb [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+            timezone: 'America/Nome'
+          })
+          getter.perform
+          records = getter.records
+          count = getter.count
+
+          assert records.count == 15
+          assert count = 30
+          assert records.first.id == 30
+          assert records.last.id == 16
```
Comment (larcin) [inline spec/services/forest_liana/resources_getter_spec.rb:78]: [TESTS] You should use rspec matchers (see example on other services tests: `spec/services/forest_liana/permissions_getter_spec.rb:44`)

File: spec/services/forest_liana/resources_getter_spec.rb [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+      describe 'when asking for the 1st page and 15 records' do
+        it 'should get only the expected records' do
+          getter = ResourcesGetter.new(User, {
+            page: { size: 15, number: 1 },
+            sort: '-id',
+            timezone: 'America/Nome'
+          })
+          getter.perform
+          records = getter.records
+          count = getter.count
```
Comment (larcin) [inline spec/services/forest_liana/resources_getter_spec.rb:73]: [MISC] This one might be a bit harder to handle, but rspec provides way to avoir duplicating all this among your tests. The best way to tackle that would to it through pair programming so I can give you tips about it :)

Ping me whenever you want to see this together :)

**Review by larcin — CHANGES_REQUESTED** [review]

File: spec/services/forest_liana/resources_getter_spec.rb [hunk 6]
```diff
@@ -1,54 +1,46 @@
...
-      tree3 = Tree.create(name: 'Apple Tree', created_at: Time.now - 5.years, island: island2, owner: user3, cutter: user1)
-      tree4 = Tree.create(name: 'Pear Tree', created_at: Time.now + 4.hours, island: island4, owner: user4, cutter: user2)
-      tree5 = Tree.create(name: 'Choco Tree', created_at: Time.now, island: island4, owner: user5, cutter: user2)
-
-      location1 = Location.create(coordinates: '12345', island: island1)
-      location2 = Location.create(coordinates: '54321', island: island2)
-      location3 = Location.create(coordinates: '43215', island: island3)
-      location4 = Location.create(coordinates: '21543', island: island4)
-      location5 = Location.create(coordinates: '32154', island: island5)
+      users = Array['Michel', 'Robert', 'Vince', 'Sandro', 'Olesya', 'Romain', 'Valentin', 'Jason', 'Arnaud', 'Jeff', 'Steve', 'Marc', 'Xavier', 'Paul', 'Mickael', 'Mike', 'Maxime', 'Gertrude', 'Monique', 'Mia', 'Rachid', 'Edouard', 'Sacha', 'Caro', 'Amand', 'Nathan', 'Noémie', 'Robin', 'Gaelle', 'Isabelle']
```
Comment (larcin) [inline spec/services/forest_liana/resources_getter_spec.rb:18]: [MISC] Is the `Array` here really mandatory? 🤔

**Review by larcin — APPROVED** [review]

**Pull request merged**